File: Problem_2 - 2.py
"""
Time Complexity : O(n). 
Space Complexity : O(n).
    
"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.index = 0 
        self.hash_ = {}
        
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not preorder:
            return None 
    
        for i,j in enumerate(inorder):
            self.hash_[j] = i
        
        def helper(preorder, s, e):
            if s > e: return None
            rootval = preorder[self.index]
            self.index += 1 
            
            node = TreeNode(rootval)
            idx = self.hash_[rootval]
            
            node.left = helper(preorder, s, idx - 1)
            node.right = helper(preorder, idx + 1, e)
            return node
            
        return helper(preorder, 0, len(inorder) - 1)
        

            

File: src/Problem2_2.java
//Time Complexity: O(n)
//Space Complexity: O(h) + O(n), where h is the height of the tree.
//Code run successfully on LeetCode.

public class Problem2_2 {

    HashMap<Integer, Integer> map = new HashMap<>();
    int idx = 0;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        
        if(preorder == null || preorder.length ==0 || inorder == null || inorder.length == 0)
            return null;
        
        for(int i =0; i < inorder.length; i++){
            map.put(inorder[i], i);
        }
        
        return helper(preorder, inorder, 0, preorder.length - 1);
    }
    
    private TreeNode helper(int[] preorder, int[] inorder, int start, int end){
        
        if(start > end)
            return null;
        
        int rootVal = preorder[idx];
        idx++;
        TreeNode root = new TreeNode(rootVal);
        int rootIdx = map.get(rootVal);
        
        root.left = helper(preorder, inorder, start, rootIdx -1);
        root.right = helper(preorder, inorder, rootIdx +1, end);
        
        return root;
    }
}
